service/history/outbound_queue_factory.go

```diff
@@ -86,19 +106,27 @@ func NewOutboundQueueFactory(params outboundQueueFactoryParams) QueueFactory {
 			int64(params.Config.OutboundQueueMaxReaderCount()),
 		),
 		hostScheduler: &queues.CommonSchedulerWrapper{
-			Scheduler: ctasks.NewGroupByScheduler[queues.StateMachineTaskTypeNamespaceIDAndDestination, queues.Executable](
-				ctasks.GroupBySchedulerOptions[queues.StateMachineTaskTypeNamespaceIDAndDestination, queues.Executable]{
+			Scheduler: ctasks.NewGroupByScheduler(
+				ctasks.GroupBySchedulerOptions[
+					queues.StateMachineTaskTypeNamespaceIDAndDestination,
+					queues.Executable,
+				]{
 					Logger: params.Logger,
 					KeyFn: func(e queues.Executable) queues.StateMachineTaskTypeNamespaceIDAndDestination {
 						return grouper.KeyTyped(e.GetTask())
 					},
 					RunnableFactory: func(e queues.Executable) ctasks.Runnable {
+						key := grouper.KeyTyped(e.GetTask())
 						return ctasks.RateLimitedTaskRunnable{
-							Limiter:  rateLimiterPool.Get(grouper.KeyTyped(e.GetTask())),
-							Runnable: ctasks.RunnableTask{Task: e},
+							Limiter: rateLimiterPool.Get(key),
+							Runnable: ctasks.RunnableTask{
+								Task: queues.NewCircuitBreakerExecutable(e, circuitBreakerPool.Get(key)),
+							},
```

Comment:
We want to apply the rate limiter after the circuit breaker to get immediate rejection of tasks but with the Runnable abstraction it's not possible.
Maybe add a `queues.RateLimitedExecutable` and delete `RateLimitedTaskRunnable`?